### PR TITLE
linux-wallpaperengine: init at 0-unstable-2024-10-13

### LINKED ISSUE #348135: Package request: Re-add linux-wallpaperengine
`linux-wallpaperengine` was removed due to depending on an unmaintained library with lots of critical CVEs (https://github.com/NixOS/nixpkgs/pull/323821; https://github.com/Almamu/linux-wallpaperengine/issues/204). Now that dependency has been replaced with one that's actually secure, it should be re-added.

### PR DESCRIPTION
**Pull request by ghost**
- Init linux-wallpaperengine, Wallpaper Engine backgrounds for Linux! https://github.com/Almamu/linux-wallpaperengine
- Fixes #348135
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Noodlez1232** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 348336`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>linux-wallpaperengine</li>
  </ul>
</details>

**Comment by Noodlez1232** [comment]
I'm so glad this exists as well. I loved wallpaper engine when I was still using Windows.

**Comment by g-libardi** [comment]
Hello, I apologize if this isn’t the appropriate place, but I couldn’t find a discussion regarding Wallpaper Engine. If anyone is interested in using the version from the pull request before it's approved, you can do so with the following command:

`nix profile install github:NixOS/nixpkgs/563f07d52d6ec4a48f1e0fa008194d00d99ea2cf#linux-wallpaperengine`

This works with the experimental Nix CLI, and the link should also function within Nix files when used with the appropriate function.

**Comment by Steinhagen** [comment]
This doesn't seem to work for me:

```
# nix run github:aucub/nixpkgs/linux-wallpaperengine#linux-wallpaperengine
[1021/221813.686260:WARNING:resource_util.cc(94)] Please customize CefSettings.root_cache_path for your application. Use of the default value may lead to unintended process singleton behavior.
Cannot find directory for steam app wallpaper_engine: assets
terminate called after throwing an instance of 'WallpaperEngine::Assets::CAssetLoadException'
  what():  Cannot find file : Cannot find directory
src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000

src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000

src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000

src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000

src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000

[0100/000000.976815:WARNING:sandbox_linux.cc(400)] InitializeSandbox() called with multiple threads in process gpu-process.
zsh: IOT instruction (core dumped)  nix run github:aucub/nixpkgs/linux-wallpaperengine#linux-wallpaperengine

```

**Comment by g-libardi** [comment]
> This doesn't seem to work for me:
> 
> ```
> # nix run github:aucub/nixpkgs/linux-wallpaperengine#linux-wallpaperengine
> [1021/221813.686260:WARNING:resource_util.cc(94)] Please customize CefSettings.root_cache_path for your application. Use of the default value may lead to unintended process singleton behavior.
> Cannot find directory for steam app wallpaper_engine: assets
> terminate called after throwing an instance of 'WallpaperEngine::Assets::CAssetLoadException'
>   what():  Cannot find file : Cannot find directory
> src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> 
> src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> 
> src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> 
> src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> 
> src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> 
> [0100/000000.976815:WARNING:sandbox_linux.cc(400)] InitializeSandbox() called with multiple threads in process gpu-process.
> zsh: IOT instruction (core dumped)  nix run github:aucub/nixpkgs/linux-wallpaperengine#linux-wallpaperengine
> ```

@Steinhagen Actually i do receive this issue too every time i run the pkg, but it works just fine. The problem is that you are just running it without passing any args. Try installing it with `nix profile install` just like i did and then run it as:
`linux-wallpaperengine <workshop_id>`

I know this one from frieren works for sure:
https://steamcommunity.com/sharedfiles/filedetails/?id=3135696887
Be sure to subscribe to it

**Comment by Steinhagen** [comment]
This still doesn't seem to work with: `linux-wallpaperengine <workshop_id>`

```
# linux-wallpaperengine 3135696887
[1021/223143.093652:WARNING:resource_util.cc(94)] Please customize CefSettings.root_cache_path for your application. Use of the default value may lead to unintended process singleton behavior.
Cannot find workshop directory for steam app 431960 and content 3135696887
terminate called after throwing an instance of 'std::runtime_error'
  what():  Cannot find workshop directory for steam app 431960 and content 3135696887
src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000

src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000

src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000

src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000

src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000

[0100/000000.413098:WARNING:sandbox_linux.cc(400)] InitializeSandbox() called with multiple threads in process gpu-process.
zsh: IOT instruction (core dumped)  linux-wallpaperengine 3135696887
```

**Comment by aucub** [comment]
> This still doesn't seem to work with: `linux-wallpaperengine <workshop_id>`
> 
> ```
> # linux-wallpaperengine 3135696887
> [1021/223143.093652:WARNING:resource_util.cc(94)] Please customize CefSettings.root_cache_path for your application. Use of the default value may lead to unintended process singleton behavior.
> Cannot find workshop directory for steam app 431960 and content 3135696887
> terminate called after throwing an instance of 'std::runtime_error'
>   what():  Cannot find workshop directory for steam app 431960 and content 3135696887
> src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> 
> src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> 
> src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> 
> src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> 
> src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> 
> [0100/000000.413098:WARNING:sandbox_linux.cc(400)] InitializeSandbox() called with multiple threads in process gpu-process.
> zsh: IOT instruction (core dumped)  linux-wallpaperengine 3135696887
> ```

Is steam and wallpaper engine installed?

**Comment by Steinhagen** [comment]
> > This still doesn't seem to work with: `linux-wallpaperengine <workshop_id>`
> > ```
> > # linux-wallpaperengine 3135696887
> > [1021/223143.093652:WARNING:resource_util.cc(94)] Please customize CefSettings.root_cache_path for your application. Use of the default value may lead to unintended process singleton behavior.
> > Cannot find workshop directory for steam app 431960 and content 3135696887
> > terminate called after throwing an instance of 'std::runtime_error'
> >   what():  Cannot find workshop directory for steam app 431960 and content 3135696887
> > src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> > 
> > src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> > 
> > src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> > 
> > src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> > 
> > src/gbm_drv_common.c:131: GBM-DRV error (get_bytes_per_component): Unknown or not supported format: 808530000
> > 
> > [0100/000000.413098:WARNING:sandbox_linux.cc(400)] InitializeSandbox() called with multiple threads in process gpu-process.
> > zsh: IOT instruction (core dumped)  linux-wallpaperengine 3135696887
> > ```
> 
> Is steam and wallpaper engine installed?

I was under the impression that this is now a standalone app and doesn't rely on steam anymore (I haven't used the app in a few years).
I will correctly setup up wallpaperengine on steam and give this another try. Thanks and sorry for the confusion.

**Comment by Steinhagen** [comment]
@aucub Maybe something that I'm messing up, but do you know why this is failing?

image

P.S.: The app was installed using: `nix shell github:aucub/nixpkgs/linux-wallpaperengine#linux-wallpaperengine`

**Comment by aucub** [comment]
> @aucub Maybe it's something that I'm messing up, but do you know why this is failing?
> 
> image
> 
> P.S.: The app was installed using: `nix shell github:aucub/nixpkgs/linux-wallpaperengine#linux-wallpaperengine`

You can use the `--assets-dir` parameter to specify the assets directory, and specify the background directory with `--bg`.

**Comment by Steinhagen** [comment]
After properly starting the wallpaper engine and subscribing to the desired asset, `linux-wallpaperengine <workshop_id>` works as expected. This was a problem with the way I was using it.
Other than that, the code looks good and the vulnerability was removed. LGTM.

**Review by Steinhagen — APPROVED** [review]

File: pkgs/by-name/li/linux-wallpaperengine/package.nix [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    wayland
+    wayland-protocols
+    egl-wayland
+    libffi
+    wayland-scanner
+  ];
+  rpath = lib.makeLibraryPath buildInputs;
+in
+stdenv.mkDerivation {
+  inherit buildInputs;
```
Comment (SuperSandro2000) [inline pkgs/by-name/li/linux-wallpaperengine/package.nix:149]:
```suggestion
```

File: pkgs/by-name/li/linux-wallpaperengine/package.nix [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    repo = "linux-wallpaperengine";
+    rev = "ec60a8a57153e49e3684c864a6d809fe9601336b";
+    hash = "sha256-M77Wp6tCXO2oFgfZ0+mdBT07CCYLsDDyHjeHtaDVvu8=";
+  };
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+    autoPatchelfHook
+  ];
```
Comment (SuperSandro2000) [inline pkgs/by-name/li/linux-wallpaperengine/package.nix:164]:
```suggestion
  ];

  inherit buildInputs;

```

File: pkgs/by-name/li/linux-wallpaperengine/package.nix [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE=${buildType}"
+    "-DCMAKE_INSTALL_PREFIX=${placeholder "out"}/linux-wallpaperengine"
+  ];
+
+  preConfigure = ''
+    patchShebangs .
+    mkdir -p third_party/cef/
+    ln -s ${cef-bin} third_party/cef/${cef-bin-name}
+  '';
```
Comment (SuperSandro2000) [inline pkgs/by-name/li/linux-wallpaperengine/package.nix:175]: This should probably all be done in postPatch

File: pkgs/by-name/li/linux-wallpaperengine/package.nix [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+      chmod +w -R $out/
+      patchelf $out/${buildType}/libcef.so --set-rpath "${rpath}" --add-needed libudev.so
+      patchelf $out/${buildType}/libGLESv2.so --set-rpath "${rpath}" --add-needed libGL.so.1
+      patchelf $out/${buildType}/chrome-sandbox --set-interpreter $(cat $NIX_BINTOOLS/nix-support/dynamic-linker)
+      sed 's/-O0/-O2/' -i $out/cmake/cef_variables.cmake
+
+      runHook postInstall
+    '';
+  };
+  buildInputs = [
```
Comment (SuperSandro2000) [inline pkgs/by-name/li/linux-wallpaperengine/package.nix:97]: Does cef really need all of this?

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**